dhall/src/Dhall/Schemas.hs

```diff
@@ -0,0 +1,206 @@
+decodeSchema _ =
+    Nothing
+
+decodeSchemas
+    :: Expr Src Void
+    -> Maybe (Map SHA256Digest (Text, Map Text (Expr Src Void)))
+decodeSchemas (RecordLit keyValues) = do
+    m <- traverse decodeSchema keyValues
+
+    let typeMetadata = Map.fromList $ do
+            (name, (_Type, _default)) <- Map.toList m
+
+            return (Import.hashExpression (Syntax.denote _Type), (name, _default))
+
+    return typeMetadata
+decodeSchemas  _ = do
+    empty
+
+-- | Simplify a Dhall expression using a record of schemas
+simplifyUsingSchemas
+    :: Expr Src Import
+    -- ^ Record of schemas
+    -> Expr Src Import
+    -- ^ Expression to simplify using the supplied schemas
+    -> IO (Expr Src Import)
+simplifyUsingSchemas _schemas expression = do
+    resolvedSchemas    <- Import.load _schemas
+    resolvedExpression <- Import.load expression
+
+    _ <- Core.throws (TypeCheck.typeOf resolvedSchemas)
+    _ <- Core.throws (TypeCheck.typeOf resolvedExpression)
+
+    let normalizedSchemas    = Normalize.normalize resolvedSchemas
+    let normalizedExpression = Normalize.normalize resolvedExpression
+
+    typeMetadata <- case decodeSchemas normalizedSchemas of
+        Just typeMetadata -> return typeMetadata
+        Nothing           -> Exception.throwIO NotASchemaRecord
+
+    let schemasRewrite subExpression@(RecordLit keyValues) =
+            Maybe.fromMaybe subExpression $ do
+                let substitutions = Map.singleton "schemas" normalizedSchemas
+
+                let substitutedExpression =
+                        Substitution.substitute (Syntax.denote subExpression) substitutions
+
+                hash <- case TypeCheck.typeOf substitutedExpression of
+                    Left _ ->
+                        empty
+                    Right subExpressionType ->
+                        return (Import.hashExpression subExpressionType)
+
+                (name, _default) <- Map.lookup hash typeMetadata
+
+                let diff a b | a == b    = Nothing
+                             | otherwise = Just a
+
+                let defaultedKeyValues =
+                        Map.fromMap (Data.Map.differenceWith diff (Map.toMap keyValues) (Map.toMap _default))
+
+                let defaultedRecord = RecordLit defaultedKeyValues
+
+                return (RecordCompletion (Field "schemas" name) defaultedRecord)
+        schemasRewrite subExpression =
+            subExpression
+
+    let rewrittenExpression =
+            fmap Void.absurd (Optics.transformOf Syntax.subExpressions schemasRewrite normalizedExpression)
+
+    let hasSchemas (Var (V "schemas" 0)) = True
+        hasSchemas _                     = False
+
+    let usesSchema =
+            Lens.anyOf
+                (Optics.cosmosOf Syntax.subExpressions)
+                hasSchemas
+                rewrittenExpression
+
+    if usesSchema
+        then return (Let (Syntax.makeBinding "schemas" _schemas) rewrittenExpression)
+        else return rewrittenExpression
```

Comment:
Is there a reason why we ought to return the normalized expression, even when it doesn't use the schemas?

```suggestion
        else return expression
```

Reply:
Alternatively, `simplifyUsingSchemas` could return a `Maybe` to signal whether the schemas are used.